### PR TITLE
DEPR: Deprecate set and dict as indexers

### LINKED ISSUE #42825: DOC: .loc doesn't document that it can take sets or dicts as arguments
#### Location of the documentation

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html?highlight=loc#pandas.DataFrame.loc
#### Documentation problem

Surprisingly, the following code works:
```python
import pandas as pd

df = pd.DataFrame([["a", 1], ["b", 2], ["c", 3]], columns=["let", "num"]).set_index(
    "let"
)

dfl = df.loc[set(["a", "b"]), :]

print(dfl)

dd = {"a": 10, "c": 12}

dfd = df.loc[dd, "num"]

print(dfd)
```

In the documentation, we don't say that you can pass a `set` or a `dict` as one of the arguments to `.loc` .

So either we update the documentation, or maybe we should be raising an Exception?

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #42825
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

- ``DataFrame.__setitem__`` does not work for unhashable object
- ``iloc`` does not work for unhashable objects for __getitem__ and __setitem__

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3456,7 +3456,32 @@ def _iter_column_arrays(self) -> Iterator[ArrayLike]:
         for i in range(len(self.columns)):
             yield self._get_column_array(i)
 
+    def _check_deprecated_indexers(self, key):
```
Comment (jreback) [inline pandas/core/frame.py:3459]: can this be a function in `indexing.py`  or similar?

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -942,7 +950,32 @@ def _getitem_nested_tuple(self, tup: tuple):
     def _convert_to_indexer(self, key, axis: int):
         raise AbstractMethodError(self)
 
+    def _check_deprecated_indexers(self, key):
```
Comment (jreback) [inline pandas/core/indexing.py:953]: same make this a function (and just call it here)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 1 above)
Comment (phofl) [reply pandas/core/frame.py]: Yep done, hadn't had much time on friday and wanted to check at least whether I got every occurrence in the test suite

**Review by mroeschke — COMMENTED** [review]
> iloc does not work for unhashable objects for getitem and setitem

Do we have tests for this behavior? Otherwise LGTM

**Comment by phofl** [comment]
Not 100% sure, but wanted to add explicit checks as followups to raise a controlled error. Something we can reuse for the deprecated cases.

**Comment by jreback** [comment]
> Not 100% sure, but wanted to add explicit checks as followups to raise a controlled error. Something we can reuse for the deprecated cases.

sgtm

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**